kratos/spaces/ublas_space.h

```diff
@@ -689,25 +679,174 @@ class UblasSpace
         }
     }
 
+    /**
+     * @brief This method checks and corrects the zero diagonal values
+     * @details This method returns the scale norm considering for scaling the diagonal
+     * @param rProcessInfo The problem process info
+     * @param rA The LHS matrix
+     * @param rb The RHS vector
+     * @param ScalingDiagonal The type of caling diagonal considered
+     * @return The scale norm
+     */
+    static double CheckAndCorrectZeroDiagonalValues(
+        const ProcessInfo& rProcessInfo,
+        MatrixType& rA,
+        VectorType& rb,
+        const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING
+        )
+    {
+        const std::size_t system_size = rA.size1();
+
+        const double* Avalues = rA.value_data().begin();
+        const std::size_t* Arow_indices = rA.index1_data().begin();
+
+        // Define  zero value tolerance
+        const double zero_tolerance = std::numeric_limits<double>::epsilon();
+
+        // The diagonal considered
+        const double scale_factor = GetScaleNorm(rProcessInfo, rA, ScalingDiagonal);
+
+        // Detect if there is a line of all zeros and set the diagonal to a 1 if this happens
+        IndexPartition<std::size_t>(system_size).for_each([&](std::size_t Index){
```

Comment:
```suggestion
        IndexPartition(system_size).for_each([&](std::size_t Index){
```

FYI this is possible with C++17
Class template argument deduction (CTAD)

Reply:
![](https://media.tenor.com/ioWF6s2tuD4AAAAM/spongegar-funny.gif)